libraries/AP_GPS/AP_GPS.cpp

```diff
@@ -700,3 +705,88 @@ AP_GPS::_broadcast_gps_type(const char *type, uint8_t instance, int8_t baud_inde
     }
     GCS_MAVLINK::send_statustext_all(MAV_SEVERITY_INFO, buffer);
 }
+
+
+/* 
+   re-assemble GPS_RTCM_DATA message
+ */
+void AP_GPS::handle_gps_rtcm_data(const mavlink_message_t *msg)
+{
+    mavlink_gps_rtcm_data_t packet;
+    mavlink_msg_gps_rtcm_data_decode(msg, &packet);
+
+    if (packet.len > MAVLINK_MSG_GPS_RTCM_DATA_FIELD_DATA_LEN) {
+        // invalid packet
+        return;
+    }
+    
+    if ((packet.flags & 1) == 0) {
+        // it is not fragmented, pass direct
+        inject_data_all(packet.data, packet.len);
+        return;
+    }
+
+    // see if we need to allocate re-assembly buffer
+    if (rtcm_buffer == nullptr) {
+        rtcm_buffer = (struct rtcm_buffer *)calloc(1, sizeof(*rtcm_buffer));
+        if (rtcm_buffer == nullptr) {
+            // nothing to do but discard the data
+            return;
+        }
+    }
+
+    uint8_t fragment = (packet.flags >> 1U) & 0x03;
+    uint8_t sequence = (packet.flags >> 3U) & 0x1F;
+
+    // see if this fragment is consistent with existing fragments
+    if (rtcm_buffer->fragments_received &&
+        (rtcm_buffer->sequence != sequence ||
+         rtcm_buffer->fragments_received & (1U<<fragment))) {
```

Comment:
> However if they send a new fragment with the same sequence & fragment id but different data, that is clearly not compatible and we should drop the buffer in favor of this. (So memcmp makes sense)

My only issue with this is: how do you know which one is the correct one? You have already received a fragment from a specific sequence, then you receive the same fragment, but the content is different, how do you know if which one is correct?
That is why I think we should just ignore a fragment we have already received, because the case you mention is an incorrect implementation of the protocol in the GCS side and fixing it on our side is just playing guesses.